### PR TITLE
GatherND-12 Implementation

### PR DESCRIPTION
**Pull request by SherlockNoMad**

### CODE REVIEW

File: onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc [hunk 1]
```diff
@@ -48,7 +57,9 @@ TEST(GatherNDOpTest, string) {
   RunTest<std::string>({2, 2}, {"ab", "cde", "f", "ghi"}, {2, 1, 1}, {1LL, 0LL}, {2, 1, 2}, {"f", "ghi", "ab", "cde"});
 
   // with negative indices
-  RunTest<std::string>({2, 2}, {"ab", "cde", "f", "ghi"}, {2, 1, 1}, {-1, 0}, {2, 1, 2}, {"f", "ghi", "ab", "cde"});
+  
+  // Fix this
+  // RunTest<std::string>({2, 2}, {"ab", "cde", "f", "ghi"}, {2, 1, 1}, {-1, 0}, {2, 1, 2}, {"f", "ghi", "ab", "cde"});
```
Comment (SherlockNoMad) [inline onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc:62]: Fix this

File: onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc [hunk 2]
```diff
@@ -65,8 +76,10 @@ TEST(GatherNDOpTest, int64_t) {
                    {4LL, 5LL, 6LL, 7LL, 0LL, 1LL, 2LL, 3LL});
 
   // with negative indices
-  RunTest<int64_t>({2, 2, 2}, {0LL, 1LL, 2LL, 3LL, 4LL, 5LL, 6LL, 7LL}, {2, 1, 1}, {-1, 0}, {2, 1, 2, 2},
-                   {4LL, 5LL, 6LL, 7LL, 0LL, 1LL, 2LL, 3LL});
+
+  // Fix this
+  // RunTest<int64_t>({2, 2, 2}, {0LL, 1LL, 2LL, 3LL, 4LL, 5LL, 6LL, 7LL}, {2, 1, 1}, {-1, 0}, {2, 1, 2, 2},
```
Comment (SherlockNoMad) [inline onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc:81]: Fix this

File: onnxruntime/core/providers/cuda/tensor/gather_nd_impl.cu [hunk 3]
```diff
@@ -26,7 +28,12 @@ __global__ void _ComputeSliceOffsetsKernel(
...
   size_t relative_slice_offset = 0;
   for (size_t dim_idx = 0; dim_idx < num_slice_dims; ++dim_idx) {
-    relative_slice_offset += static_cast<int64_t>(slice_indices[dim_idx]) * sizes_from_slice_dims_data[dim_idx];
+    int64_t index = static_cast<int64_t>(slice_indices[dim_idx]);
+    const size_t input_dim_idx = batch_dims + dim_idx;
+    CUDA_KERNEL_ASSERT(index >= -input_dims[input_dim_idx] && index < input_dims[input_dim_idx]);
+    if (index < 0) index += input_dims[input_dim_idx];
+
+    relative_slice_offset += index * sizes_from_slice_dims_data[dim_idx];
   }
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/tensor/gather_nd_impl.cu:37]: This part is new code

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 4]
```diff
@@ -32,6 +32,14 @@ ONNX_CPU_OPERATOR_KERNEL(GatherND, 11,
                              .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
                          GatherND);
 
+ONNX_CPU_OPERATOR_KERNEL(
```
Comment (pengwa) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc:35]: nit: looks this format is different with the above opset 11 registry.

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 5]
```diff
@@ -107,11 +107,16 @@ Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) con
       reminder -= (idx * element_counts[last_indices_dimension + j]);
     }
     for (int64_t j = batch_dims_; j < last_indices_dimension; ++j) {
-      auto indice = *(indice_offset + i * (last_indices_dimension - batch_dims_) + (j - batch_dims_));
-      if (indice < 0 || indice >= input_shape[j]) {
-        err_index = indice;
+      auto index = *(indice_offset + i * (last_indices_dimension - batch_dims_) + (j - batch_dims_));
+      auto upper_limit = input_shape[j];
+      auto lower_limit = -upper_limit;
+      if (index < lower_limit || index >= input_shape[j]) {
```
Comment (pengwa) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc]: nit: change to index >= upper_limit?

File: onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc [hunk 6]
```diff
@@ -55,11 +55,9 @@ TEST(GatherNDOpTest, string) {
                        {2, 1, 1, 0, 0, 1}, {3}, {"kite", "bob", "dance"});
 
   RunTest<std::string>({2, 2}, {"ab", "cde", "f", "ghi"}, {2, 1, 1}, {1LL, 0LL}, {2, 1, 2}, {"f", "ghi", "ab", "cde"});
-
-  // with negative indices
   
-  // Fix this
-  // RunTest<std::string>({2, 2}, {"ab", "cde", "f", "ghi"}, {2, 1, 1}, {-1, 0}, {2, 1, 2}, {"f", "ghi", "ab", "cde"});
+  // with negative indices
```
Comment (pengwa) [inline onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc]: shall we have a batch_dim != 0 with negative indice case?

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 7]
```diff
@@ -63,12 +71,20 @@ Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) con
...
+  }
+
   int64_t err_index = 0;
   p.element_bytes = input_tensor->DataType()->Size();
   p.element_to_copy = input_shape.SizeFromDimension(last_indices_dimension);
   p.bytes_to_copy = p.element_bytes * p.element_to_copy;
-  const auto* indices_data = indices_tensor->Data<Tind>();
-  const int64_t offset_count = indices_shape.Size() / last_indices_dimension;  // Times to copy
+  const auto* indice_offset = indices_tensor->Data<Tind>();
+  const int64_t offset_count = indices_shape.Size() / (last_indices_dimension - batch_dims_);  // Times to copy
```
Comment (pengwa) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc:87]: I found the naming in this function is pretty confusing, (hard to understand), I can help rename them in my later PR.

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 8]
```diff
@@ -63,12 +71,20 @@ Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) con
     element_counts[i] = input_shape.SizeFromDimension(i + 1);
   }
 
+  auto last_dim_size = indices_shape.SizeFromDimension(indices_shape.NumDimensions() - 1);
```
Comment (pengwa) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc:74]: is it same as "auto last_dim_size = indice_shape[indice_shape.NumDimensions() - 1]"?

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc (hunk 7 above)
Comment (pengwa) [reply onnxruntime/core/providers/cpu/tensor/gather_nd.cc:87]: for example I thin calling slice_count is better than offset_count. 

element_counts[0:last_indices_dimension] is very different with element_counts[last_indices_dimension + batch_dims_ - 1:last_indices_dimension ]

Comment (SherlockNoMad) [reply onnxruntime/core/providers/cpu/tensor/gather_nd.cc:89]: Yes. Please... 
Thanks for the help. 
I will leave the name untouched in this PR

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 9]
```diff
@@ -25,12 +25,22 @@ ONNX_OPERATOR_KERNEL_EX(GatherND, kMSDomain, 1, kCpuExecutionProvider,
...
+        // official ONNX spec only supports `int64_t` for indices
+        .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
+    GatherND);
+
+ONNX_CPU_OPERATOR_KERNEL(
+    GatherND,
+    12,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
```
Comment (souptc) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc:42]: do we need int32?

Comment (SherlockNoMad) [reply onnxruntime/core/providers/cpu/tensor/gather_nd.cc:42]: No... Spec doesn't have int32

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc (hunk 8 above)
Comment (pengwa) [reply onnxruntime/core/providers/cpu/tensor/gather_nd.cc:76]: I will make the change in another PR.

**Comment by pengwa** [comment]
For left comments, I'll take care in separate PR. Merge this to unblock subsequent GradNdGrad CPU development.

**Pull request merged**